### ISSUE TITLE
Propose migration strategy

### ISSUE THREAD

**Issue by valiafetisov** [powerhouse-inc/switchboard-boilerplate#60]
### Goal
Migration strategy for the currently running API

### Context
Currently running API will need to be slowly migrated to the new backend. We should expect and plan that process to happen step-by-step, or in other words, model-by-model. This issue should outline how this is possible and potentially outline different scenarios to choose from (if it's not straight-forward).

Desired flow: make existing queries available under a new server, based on configuration. Open questions are:
- Mechanics (should graphql federation be used?)
- Authorization (should old endpoints be protected by the new authorization, should we use a single API key to protect writes, etc)
- If currently running codebase needs to be changed in any way to streamline migration
- If any new features are required for the migration (eg separation between read/write tokens)

### Links
- Existing graphql server that needs to be migrated https://studio.apollographql.com/public/Performance-Dashboard-Prd/variant/current/home
- Existing graphql API explorer https://studio.apollographql.com/public/Performance-Dashboard-Prd/variant/current/explorer
- Existing codebase https://github.com/makerdao-ses/ecosystem-api

### Tasks
- [x] Research graphql migration strategies
- [x] Outline different possibilities applicable to our situation
- [ ] Potentially work on the first implementation

**Comment by KirillDogadin-std** [comment]
FYI:

fetching the external schema does not seem straightforward.


1. I've had an error

```
rover graph introspect https://ecosystem-dashboard.herokuapp.com/graphql

error[E004]: HTTP status client error (400 Bad Request) for url (https://ecosystem-dashboard.herokua
pp.com/graphql)
        Make sure the endpoint is accepting connections and is spelled correctly

```

while doing `npm run dev` in our api service directory and running 

```
npx rover graph introspect http://localhost:3001/graphql
```

prints the schema

2. which lead to conclusion that maybe the introspection is not enabled on the ecosystem api website. Looking at the [src](https://github.com/makerdao-ses/ecosystem-api/blob/616a10d34a735a275f11649681700e648c117b26/src/index.ts#L42-L64) code i don't see the flag being `true` - and that might be the problem because as per [docs](https://www.apollographql.com/docs/apollo-server/api/apollo-server/) introspection is disabled by default in the `production` node env.

**Comment by valiafetisov** [comment]
> i don't see the flag being `true` - and that might be the problem

You can propose this to be changed, as outlined in the issue `If currently running codebase needs to be changed in any way to streamline migration` feel free to collect instructions or even open a PR (maybe first to the forked repo)

> UNKNOWN: Type of field "CoreUnit.id" is incompatible across subgraphs: it has type "ID!" in subgraph "ecosystem" b
ut type "String" in subgraph "switchboard"

That's exactly the reason why we need to come up with migration strategy: the tool currently seems to try merging those two as they match the name, but expected result would be to "ignore" the table in the old codebase and replace it with the new, right? Maybe there is a `rover` flag/config for that?

**Comment by KirillDogadin-std** [comment]
> You can propose this to be changed, as outlined in the issue If currently running codebase needs to be changed in any way to streamline migration feel free to collect instructions or even open a PR (maybe first to the forked repo) 

Will post the comment right after this one with the summary of this that would be edited over time should the new requirements appear.

> the tool currently seems to try merging those two as they match the name

Actually i don't believe "merge" to be an accurate word. More like "unite". It just refuses to create a supergraph (union) with the conflicts among the subgraphs' defenitions. Here's the [page](https://www.apollographql.com/docs/federation/federated-types/composition/#breaking-composition) with a few examples on such problems/conflicts. 

On top of this the supergraph for the federation should be created from the files which allow for this action: i.e. graphql schemas that are subgraphs. [Here](https://www.apollographql.com/docs/apollo-server/using-federation/apollo-subgraph-setup/) is the example with transformation of a usual gql schema in to a federation subgraph.

> but expected result would be to "ignore" the table in the old codebase and replace it with the new, right?

Well, i sure did not interpret the header-comment's contents such as "Currently running API will need to ... model-by-model." this way. To clear misunderstanding this we mb should have a sync-discussion. 

And regarding replacing the models with our models gradually: it seems even more like requiring manual work for schema adjustments. I would vouch for uniting the gql schemas ("how" is tbd) and deprecating the endpoints that use outdated gql (i.e. return something like `308` with the informative error message from the deprecated endpoint).

Log of looking into [this](https://github.com/powerhouse-inc/switchboard-boilerplate/issues/60#issuecomment-1647853700)
---

[Problem](https://github.com/powerhouse-inc/switchboard-boilerplate/issues/60#issuecomment-1647853700): there're multiple places where the names of the types in the
schemas collide - it becomes impossible to successfully run the rover
supergraph generation.

1.  prefix all the types with the identifier of the namespace is most
    primitive and straightforward
2.  <https://github.com/wundergraph/wundergraph> is the tool that allows
    to introspect and unify the apis
    -   👎 sadly it looks like one has to define the operations
        performed with the schema. This would lead to us having to
        rewrite the code.
        [link](https://github.com/wundergraph/wundergraph/blob/8cf658206efb45ecb0a52f3089f1d218ab086a2e/README.md?plain=1#L95)
3.  The federation approach where the schemas are combined together
    requires the base schemas to represent a subgraph I.e.
    <https://www.apollographql.com/docs/federation/building-supergraphs/subgraphs-overview/#choosing-a-subgraph-library>
    Here's the list of supported libs: [link](https://www.apollographql.com/docs/federation/building-supergraphs/supported-subgraphs#javascript--typescript)
    -   👎 Nexus is not on the list.
    -   👍 Should be one of the cleanest solutions there.
    -   ℹ️ Does not seem that the conflicts are possible in this
        approach and hence resolving merged types is not the thing here
4.  👍 Conversion of the gqlschema into subgraph seems possible:
    <https://www.apollographql.com/docs/apollo-server/using-federation/apollo-subgraph-setup/>
    👎 But it seems that manual adjustment of the schema is required
    there:
    -   <https://github.com/graphql-nexus/nexus/issues/148> still open
        and we can't already generate the subgraph/federation-compatible
        schema on our end due to limitations of the chosen lib.
5.  It's not stated explicitly at any place i've read through, but feels
    like properly generated subgraphs can be united into supergraph
    without the conflicts (as in (3)).
    <https://www.apollographql.com/docs/federation/federated-types/composition/#supported-methods>
    is the link that made me belive so.

#### Possible action:

1.  Decide to not "overwrite" entities of the external schema, just
    merge the schemas together.
2.  Perfix our types (and ask them to prefix theirs). Should resolve the
    conflicts

#### Actions that do not seam sensful:

1.  Manually convert the schemas into federation compatible subgraphs. Because we're not supposed to adjust the nexus schema on our end as per nexus' design (it says this in the top level comment of the generated files); and because asking the "3rd party" to adjust the schema on their end would be extra cumbersome. 
2.  Migrate to the frameworks that generate federation compatible
    schemas. Because it takes too much time.

**Comment by KirillDogadin-std** [comment]
# Further investigation

-   <https://github.com/protofire/nexus-mutual-subgraph/tree/master>
    this seems like a lot of problem-specific (non-general) code without
    documentation so i would deprioritize looking into what it does
    whatsoever.
-   <https://github.com/rjdmacedo/federated-nexus-prisma-subgraph-template>
    this indeed uses [npm package](https://www.npmjs.com/package/graphql-transform-federation)
    to produce a federation-ready schema. Although the following should
    be noted:
    -   The schema that is generated by nexus and stored in the git repo
        is not federation-ready. Instead the federation schema is
        [generated](https://github.com/rjdmacedo/federated-nexus-prisma-subgraph-template/blob/7fc3d88cfff42249e114bef154a16d3d6827285e/src/schema.ts#L32) in code and is provided during creation of the apollo server
        creation [here](https://github.com/rjdmacedo/federated-nexus-prisma-subgraph-template/blob/7fc3d88cfff42249e114bef154a16d3d6827285e/src/index.ts#L10)
    -   so in order to create a supergraph like it is done [here](https://www.apollographql.com/docs/rover/commands/supergraphs/#yaml-configuration-file)
        (with the schema file being specified) does not seem possible.
        Instead what should/could be done is avoiding files and
        providing the links to the services that are being united into
        the federtation. I.e. like
        [here](https://www.apollographql.com/docs/federation/quickstart/local-composition#1-provide-subgraph-details)
        I see the following properties of this approach:
        -   The federation is created based on the production / deployed
            services. It is less easy to handle during the development.
        -   Action is required on the side of the collaborator-api to
            also create the federation-ready schema endpoint.
        -   it is necessary to determine what federation version has to
            be used. It is not obvious from the [main
            page](https://www.npmjs.com/package/graphql-transform-federation)
            what version the output of the federation file is going to
            have.
        -   Unknown-unknowns:
            -   some problems can arise during the generation of the
                federation api.
-   [Wundergraph](https://github.com/wundergraph/wundergraph/blob/main/README.md)
    allows to unify the apis
    -   [example](https://github.com/wundergraph/wundergraph/tree/main/examples/apollo-federation)
        does not provide many explainers in the readme so the
        outstanding questions would be:
        -   does one have to define operations or can it be avoided?
            should be possible to avoid = will work without them. Tested
            manually if needed - we have a way to define the operations
            with one of the apis and interact with them in a typesafe
            way. E.g. define an operation that will address the old
            version of the api & have the typesafe code in our vue app.
        -   is this compatible with the current mechanism of how
            frontend remains typesafe? in the
            [example](https://github.com/wundergraph/wundergraph-demo/blob/1530af458c96016bb1f02255c71279fde4043eec/nextjs-frontend/pages/index.tsx#L2-L10)
            we can see that the operations can be imported from the
            generated code. So it is compatible, but introduces a hacky
            setup where we have multiple ways of fetching information
            from the server.
        -   does the current project layout have to change? Manually
            tested: initialized wundergraph in the \`api\` directory and
            ran the services successfully.
    -   seems worth making a PoC and see beter how it works. done.
        Reported answers to the questions above partially based on this
        XP.
    -   Outcome of the investigation:
        1.  It is relatively easy to setup.
        2.  If we want to use federation - it is possible to do so, but
            actions still have to be taken to transform the schemas that
            we currently have into the federation-compatible ones. this
            can be compatible (at least if nexus is used with one of the
            prvious options.)
        3.  If we don't want to use federation - wundergraph does all
            the job to make the gql endpoints available under the same
            service. The naming changes though to avoid collisions.
        4.  Eitherway (3) or (2) we will have to tweak the behaviours of
            outdated endpoints - e.g. disable them or something like
            this.
-   Wundergraph does not mess up the existing auth, here's the flow:
    -   started api
    -   started the basic wundergraph server with graphql endpoint
        enabled
    -   started frontend
    -   open browser on `localhost:3000`
    -   pressed "login with ethereum" and using metamask ended ap on
        "API Tokens" view.

## Proposal A

-   Allied api actions:
    1.  Enable the introspection in their API
        [here](https://github.com/makerdao-ses/ecosystem-api/blob/616a10d34a735a275f11649681700e648c117b26/src/index.ts#L42-L64)
    2.  Enable the conversion to federation via [this
        package](https://www.npmjs.com/package/graphql-transform-federation)
        or alternative and use it with apollo as in the [example with
        `federatedSchema`](https://github.com/rjdmacedo/federated-nexus-prisma-subgraph-template/blob/7fc3d88cfff42249e114bef154a16d3d6827285e/src/index.ts#L10)
-   Our adjustments:
    1.  Enable the conversion to federation via the package (as above)
    2.  Use the Wundergraph's automagic to unite the services into the
        federated supergraph and expose under the same api via [this
        move](https://github.com/wundergraph/wundergraph-demo/blob/1530af458c96016bb1f02255c71279fde4043eec/api/.wundergraph/wundergraph.config.ts#L26)
-   Cooperative actions:
    1.  Agree on which endpoints have to be disabled.
    2.  Allied api disables the graphql endpoints/resolvers to return
        permanent redirect response
-   Shortcoming:
    -   Transforming the schema requires in one way or another
        specifying how to transform the initial schema into the
        federation compatable subgraph. The problem here is that this
        requires continious maintainance along the way. I.e. when a
        schema changes - one has to adjust the federation config. And
        this has to be done on both ends
-   Due to the previous point it seems less desirable to choose this
    proposal over the Proposal B.

## Proposal B

-   Allied api actions:
    1.  Enable the introspection in their API
        [here](https://github.com/makerdao-ses/ecosystem-api/blob/616a10d34a735a275f11649681700e648c117b26/src/index.ts#L42-L64)
-   Our adjustments:
    1.  Enable the conversion to federation via the package (as above)
    2.  Use the Wundergraph's automagic to unite the services into the
        same api with differnet namespaces and expose under the same api
        via [this
        move](https://github.com/wundergraph/wundergraph-demo/blob/1530af458c96016bb1f02255c71279fde4043eec/api/.wundergraph/wundergraph.config.ts#L12)
-   Cooperative actions:
    1.  Agree on which endpoints have to be disabled.
    2.  Allied api disables the graphql endpoints/resolvers to return
        permanent redirect response

Short PoC here https://github.com/powerhouse-inc/switchboard-boilerplate/tree/migrate-api-wundergraph

## Proposal C

This proposal implies that the wundergraph package is not usable due to
e.g. the license. In this case one has to manually replace the basic
wundergraph's feature.

-   Allied api actions:
    1.  Enable the introspection in their API
        [here](https://github.com/makerdao-ses/ecosystem-api/blob/616a10d34a735a275f11649681700e648c117b26/src/index.ts#L42-L64)
-   Our adjustments:
    1.  Create a mechanism that replaces the names of the graphql
        entities: e.g. adds a predefined prefix to them.
    2.  use rover cli to merge the schemas (and create the
        infrastructure so that the created schema is always up to date.)
    3.  Use the schema in our api
        <https://www.apollographql.com/docs/apollo-server/using-federation/apollo-subgraph-setup/>
    4.  Use the gateway to unite the api just like
        <https://tyk.io/blog/an-introduction-to-graphql-federation/#:~:text=federation%20with%20Apollo.-,Creating%20a%20gateway,-To%20start%2C%20you>
-   Cooperative actions:
    1.  Agree on which endpoints have to be disabled.
    2.  Allied api disables the graphql endpoints/resolvers to return
        permanent redirect response
    3.  Agree on the naming convention so that there is going to be no
        overlap. E.g. book a prefix for the graphql models.

**Comment by valiafetisov** [comment]
> Instead the federation schema is
[generated](https://github.com/rjdmacedo/federated-nexus-prisma-subgraph-template/blob/7fc3d88cfff42249e114bef154a16d3d6827285e/src/schema.ts#L32) in code

This seems to do exactly what we need: provide extra resolvers on top of the existing schema, right? Eventually this can be turned into a "config" which I was talking about: where we define what goes where, exactly.

> Wundergraph does not mess up the existing auth, here's the flow:

- Where can I find PoC? Is [this](https://github.com/powerhouse-inc/switchboard-boilerplate/tree/migrate-api-wundergraph) it?
- Does this experiment implies that you've tried running `wundergraph` between `frontend` and `backend` services?
- Were you able to use our playground to fetch any queries that lies outside of our backend?
- Uniting two APIs automatically means different queries will require different authentication header? (one valid for our backend, one for the external service)
  - How can this "exclusivity" be potentially resolved?
  - For example: can authorization header for external service be set on our side? So that requests to the external service gets authorised on our side and then go there with another header?

> Proposal A
> Proposal B
> Proposal C

Could you please high-level outline what's the difference? Also, what's the preferred method based on the authorization discussion?

> Agree on which endpoints have to be disabled

What do you mean by endpoints here, queries/mutations? Why is this needed? Initially we will replace no types, but will have to provide a guide on how to do it in the future.

> Allied api disables the graphql endpoints/resolvers to return permanent redirect response

I don't think it's possible, since then our own "subgraph" requests will end up in the loop 🙂 

Note: this issue is not about automatically and magically merging two endpoints, but providing a roadmap of how to migrate one system to another, step by step. Can you please outline how will this be accomplished for the given proposals? Eg: what needs to be done when `CoreUnits` becomes part of the `switchboard-api` database?

**Comment by KirillDogadin-std** [comment]
> This seems to do exactly what we need: provide extra resolvers on top of the existing schema, right? Eventually this can be turned into a "config" which I was talking about: where we define what goes where, exactly.

not sure that we are on the same page here. Let me describe what it does with my words and we see if we're aligned.

[here](https://www.apollographql.com/docs/apollo-server/using-federation/apollo-subgraph-setup/#3-define-an-entity) the schema gets direcly edited with the `@key`. And in the provided example [here](https://github.com/rjdmacedo/federated-nexus-prisma-subgraph-template/blob/7fc3d88cfff42249e114bef154a16d3d6827285e/src/schema.ts#L32) the key field is specified via typescript. So i would not say that __just__ it provides "extra resolvers"
since we have the [resolver](https://github.com/rjdmacedo/federated-nexus-prisma-subgraph-template/blob/7fc3d88cfff42249e114bef154a16d3d6827285e/src/schema.ts#L39) and the [key field](https://github.com/rjdmacedo/federated-nexus-prisma-subgraph-template/blob/7fc3d88cfff42249e114bef154a16d3d6827285e/src/schema.ts#L38)

---

> Where can I find PoC? Is [this](https://github.com/powerhouse-inc/switchboard-boilerplate/tree/migrate-api-wundergraph) it?

yes as per this part of the post:

> Short PoC here https://github.com/powerhouse-inc/switchboard-boilerplate/tree/migrate-api-wundergraph

---

> Does this experiment implies that you've tried running wundergraph between frontend and backend services?

yes.

> Were you able to use our playground to fetch any queries that lies outside of our backend?

yes

> Uniting two APIs automatically means different queries will require different authentication header? 

yes

> How can this "exclusivity" be potentially resolved?

According to the list of wundergraph features - it's not out of the box compatible with wallet auth we have (see https://docs.wundergraph.com/docs/features/openid-connect-based-authentication). It is also hard to imagine how one can share auth without custom solution here.

> For example: can authorization header for external service be set on our side? So that requests to the external service gets authorised on our side and then go there with another header?

setting headers is possible e.g. as with [this](https://docs.wundergraph.com/docs/wundergraph-config-ts-reference/configure-headers-for-http-based-data-sources#dynamic-headers). SO it is possible to seamlessly auth the user on both servers while they login on ours and then use the linked functionality to add the header.

> Could you please high-level outline what's the difference? 

Proposal A - Wundergraph and the graphql federation
Proposal B - Wundergraph and no federation: stitching instead.
Proposal C - No wundergraph, custom workaround for the similar effect as in Proposal B.

> What do you mean by endpoints here, queries/mutations?

yes

> Why is this needed? Initially we will replace no types, but will have to provide a guide on how to do it in the future.

I assumed that from the get go it should be already replaced. Then disabling is not required.

>  Can you please outline how will this be accomplished for the given proposals? Eg: what needs to be done when CoreUnits becomes part of the switchboard-api database?

Before outlining i would first like to agree on the list of problems that occur when we have 2 APIs together (either united or federated):
  1. differences between the database contents. Some data can be present in one service's table but not in the other one.
      - in case of federation the concept of [entities](https://www.apollographql.com/docs/federation/entities) will be used and it is all fine as long as the object is present in both databases. For the case where the data is mismatched - the responses with information about the objects (e.g. Coreunits) will be incomplete.
        - Therefore the tasks at hand here would be to: 1. design the `transform-federation` configs carefully; 2. try to look for a way of having same coreunits in both dbs
      - in case of just merging/uniting the APIs they will have 0 idea about how to provide information cooperatively because federation is not in place. Consequently, the database models would have to be made identical and data has to be copied from one to another.
  2. Headers/auth have to be there.
      - this should be solvable by the wundergraph header magic (see above/way above in this post)
      - Eventually we should only rely on the single source of auth and remove this magic.

please indicate if you see any other points (e.g. No. 3) that i've missed. Please indicate if you agree or disagree with the raised bullets there.

**Comment by valiafetisov** [comment]
>> How can this "exclusivity" be potentially resolved?
>
> According to the list of wundergraph features - it's not out of the box compatible with wallet auth we have

I'm not sure I understand what you mean. On the wundergraph side the only feature we need is to pass `Authorization` header (or the cookie) to the underlying service. According to your other answer I assume it's already been tested, as you were able to login. 

But the question was not about wundergraph, but about the solution to the problem that different queries will expect different value in the `Authorization` header. The question is how it can be solved?

- One "solution: is: keep it as is, ask clients to manage authorization header depending on the query
- Second solution: require only switchboard auth for all queries even going to ecosystem API. When switchboard-valid `Authorization` header is provided, replace it with ecosystem-valid `Authorization` header
  - Would it be possible to implement something like this with wundergraph? (see related question below)

> setting headers is possible e.g. as with [this](https://docs.wundergraph.com/docs/wundergraph-config-ts-reference/configure-headers-for-http-based-data-sources#dynamic-headers)

Is `headers: (builder) => builder.addClientRequestHeader('X-Authorization', 'Authorization')` called on every request or only during schema build / on wundergraph start?

> in case of federation the concept of [entities](https://www.apollographql.com/docs/federation/entities) will be used and it is all fine as long as the object is present in both databases. For the case where the data is mismatched - the responses with information about the objects (e.g. Coreunits) will be incomplete.

Again the desirable situation is to have some kind of config on our side which clearly defines which entity to use from which service. Example situation: 
1. Default situation after migration is set up: User executes query against `switchboard-boilerplate` to get `CoreUnits` – the request is proxied to the `ecosystem-api`, 5 core units are returned
2. The data of all `CoreUnits` is successfully copied from the `ecosystem-api` to `switchboard-boilerplate` database, the "routing config" is edited
3. User executes query against `switchboard-boilerplate` to get `CoreUnits` – the request stays within `switchboard-boilerplate` and returns 5 migrated core units
4. User executes mutation against  `switchboard-boilerplate` to remove one `CoreUnit` – `switchboard-boilerplate` database is updated, `ecosystem-api` database stays the same

In other words, it's desirable to have a single source of truth (defined in code) for each entity rather then let the federation automatically decide which service to proxy the request based on the requested fields

> Headers/auth have to be there

See my solutions and questions about it above

**Comment by KirillDogadin-std** [comment]
> Is headers: (builder) => builder.addClientRequestHeader('X-Authorization', 'Authorization') called on every request or only during schema build / on wundergraph start?

Every request. Tested manually with `console.log`

> Second solution: require only switchboard auth for all queries even going to ecosystem API. When switchboard-valid Authorization header is provided, replace it with ecosystem-valid Authorization header
Would it be possible to implement something like this with wundergraph? (see related question below)

The answer below also covers the part about the config.

[here](https://docs.wundergraph.com/docs/wundergraph-server-ts-reference/custom-resolve-hook) is the description about custom resolves. It works for queries and mutations. So I would imagine that we could create a custom resolver which on "createChallenge" and "solveChallenge" mutation calls some logic that seemlessly calls login mutations on ecosystem. Or as you said replacing the autherization header with the valid ecosystem one.

That should also work for CoreUnit mutations/queries and so on.

**Comment by KirillDogadin-std** [comment]
What page?

Welcome page without functionality saying something along the lines "powered by wundergraph"

This does not seem relevant anymore due to the points you've raised below the quiestion.
